Домашнее задание Coroutines Homework

evtushenko99 opened the pull request.

Commit:
Домашнее задание Coroutines Homework

antonkazakov commented on app/src/main/java/otus/homework/coroutines/models/presentation/Text.kt:
```diff
@@ -0,0 +1,14 @@
+
+import android.content.Context
+
+sealed class Text {
```
А зачем этот класс нужен?

evtushenko99 replied:
Для более удобной работы с текстом. Можно создавать одну сущность и работать с ней во вью. Так как у нас есть ресурсы и простые строки.

antonkazakov commented on app/src/main/java/otus/homework/coroutines/presentation/CatResult.kt:
```diff
@@ -0,0 +1,11 @@
+sealed class CatResult {
+    data class Success(val catInfo: CatInfoModel) : CatResult()
+    sealed class Error : CatResult(){
+        data class ByString(val message:String):Error()
```
Тоже непонятно зачем такое решение

evtushenko99 replied:
Sealed class излишний тут

antonkazakov commented on app/src/main/java/otus/homework/coroutines/presentation/CatsPresenter.kt:
```diff
@@ -0,0 +1,56 @@
+    fun onClick() {
+        mainJob = presenterScope.launch {
+            try {
+                val fact: Deferred<CatFact> = async(Dispatchers.IO) { catsService.getCatFact() }
```
Можно не переключать диспатчер, ретрофит поддерживает саспенд функции

evtushenko99 replied:
но такие операции ведь лучше выполнять на UI потоке, а у меня в presenterScope Dispatchers.Main

antonkazakov replied:
Ответил чуть ниже

antonkazakov commented on app/src/main/java/otus/homework/coroutines/presentation/CatsPresenter.kt:
```diff
@@ -0,0 +1,56 @@
+    fun onClick() {
+        mainJob = presenterScope.launch {
+            try {
+                val fact: Deferred<CatFact> = async(Dispatchers.IO) { catsService.getCatFact() }
```
А зачем типы тут явно обьявлены?

evtushenko99 replied:
для удобства просмотра пулл реквеста

antonkazakov replied:
Спасибо)

antonkazakov commented on app/src/main/java/otus/homework/coroutines/presentation/CatsPresenter.kt:
```diff
@@ -0,0 +1,56 @@
+                _catsView?.populate(CatInfoModel(fact.await().text, icon.url, icon.width, icon.height))
+            } catch (socketException: SocketTimeoutException) {
+                _catsView?.showToast(Text.TextByRes(R.string.toast_exception))
+            } catch (ex: Exception) {
```
Тут еще нужно пробросить дальше CancellationException

evtushenko99 replied:
В задании было "Добавить обработку исключений через try-catch. В случае java.net.SocketTimeoutException показываем Toast с текстом "Не удалось получить ответ от сервером". В остальных случаях логируем исключение в otus.homework.coroutines.CrashMonitor и показываем Toast с exception.message"
SocketTimeoutException - отображаю,остальное показываю в  toast

Не очень понял, о чем речь

antonkazakov replied:
Речь о том, что ты используя catch (ex: Exception), ловишь абсолютно все исключения. CancellationException, который корутины используются для механизма отмены, также является наследником Exception, поэтому тебе нужно пробросить его дальше чтобы не сломать механизм отмены.

antonkazakov commented on app/src/main/java/otus/homework/coroutines/presentation/CatsPresenter.kt:
```diff
@@ -0,0 +1,56 @@
+
+    private var _catsView: ICatsView? = null
+
+    private val presenterScope = CoroutineScope(Dispatchers.Main + CoroutineName("CatsCoroutine"))
```
А ты не хочешь тут менять направление исключений? Если картинка не загрузится почему бы текст не отобразить?

evtushenko99 replied:
ЧТо значит менять направление исключений ?
запустить независимо загрузку иконок и загрузки текста и не ждать, пока оба загрузятся?

antonkazakov replied:
Да, тоесть если картинка например не загрузится, то показать какую то дефолтную картинку но текст показать тот что из сети пришел

antonkazakov commented on app/src/main/java/otus/homework/coroutines/presentation/MainViewModel.kt:
```diff
@@ -0,0 +1,90 @@
+    private val toastTextMutableLiveData: MutableLiveData<Text> = MutableLiveData()
+    private val catInfoMutableLiveData: MutableLiveData<CatInfoModel> = MutableLiveData()
+
+    val toastTextLiveData: LiveData<Text> = toastTextMutableLiveData
```
А зачем тебе две лайвдаты? Просто эмить какой нибудь Result.Error/Result.Success и типизируй лайвдату Result

evtushenko99 replied:
Да, Result это хорошо, но показалось излишним

antonkazakov commented on app/src/main/java/otus/homework/coroutines/presentation/MainViewModel.kt:
```diff
@@ -0,0 +1,90 @@
+    fun onInit() {
+        viewModelScope.launch(exceptionHandler) {
+
+            val fact: Deferred<CatFact> = async(Dispatchers.IO) { catsService.getCatFact() }
```
Можно не переключать диспатчер

evtushenko99 replied:
viewModelScope - Dispatcher.Main - мы же не можем на нем вызывать такие операции

antonkazakov replied:
Можем, если поддерживаются саспенд функции, они неблокирующие. Например рум и ретрофит их поддерживают. Если бы функция была блокирующая, то да

antonkazakov commented on app/src/main/java/otus/homework/coroutines/presentation/MainViewModel.kt:
```diff
@@ -0,0 +1,90 @@
+
+    private val exceptionHandler =
+        CoroutineExceptionHandler { _, throwable: Throwable ->
+            if (throwable is SocketTimeoutException) {
```
Лучше конечно через try/caatch обработать исключения, то что не поймали уже пробросится сюда и будет какой то дефолтный обработчик

evtushenko99 replied:
А почему так лучше ? Ведь exceptionHandler можно написать один раз и отдавать его всем, а тут нужно каждый раз везде try/catch

antonkazakov replied:
Основная разница между try/catch и CoroutineExceptionHandler в том, что в try/catch ты можешь сделать какой-либо ретрай, накрутить фоллбек логику и т.п., пока у тебя корутина не завершена. Если ты попал в CoroutineExceptionHandler то корутина уже в терминальном состоянии. У нас очень простой кейс, поэтому разницы здесь нет, но например если бы ты хотел после ошибки из сети пойти в БД, то тебе не подошел бы CoroutineExceptionHandler. Можно еще взять runCatching

antonkazakov commented on app/src/main/java/otus/homework/coroutines/presentation/MainViewModel.kt:
```diff
@@ -0,0 +1,90 @@
+    private fun onClickFailed(result: CatResult.Error) {
+        when (result) {
+            is CatResult.Error.ByRes -> {
+                toastTextMutableLiveData.postValue(Text.TextByRes(result.stringRes))
```
Можно взять setValue

evtushenko99 replied:
согласен

antonkazakov reviewed (changes requested):
Привет. Оставил пару комментов

Commit:
Домашнее задание Coroutines Homework часть 2

 - изменил CatResult, упростил его
 - сделал единую подписку на CatResult через LiveData
 - стал использовать методы setValue,вместо postValue у LiveData, так как мы уже на главном потоке и это излишне

Commit:
Домашнее задание Coroutines Homework часть 3

 - в CatsPresenter.kt добавил exceptionHandler, который будет ловить все  CancellationException, SupervisorJob(), который позволит не падать остальным корутинам, и пробросил дальше CancellationException, чтобы его поймал exceptionHandler
 - В презентере и вьюмоделе будет возвращаться дефолтная иконка, если пришло исключение при запросе иконки
 - Во вью модели сделал обработку исключений через try|catch
